codegen/src/main/twirl/templates/ScalaServer/Handler.scala.txt

```diff
@@ -76,11 +77,36 @@ object @{service.name}Handler {
       case m => Future.failed(new NotImplementedError(s"Not implemented: $m"))
     }
 
-    Function.unlift((req: HttpRequest) => req.uri.path match {
-      case Path.Slash(Segment(`prefix`, Path.Slash(Segment(method, Path.Empty)))) ⇒
-        Some(handle(req, method).recoverWith(GrpcExceptionHandler.default))
-      case _ =>
-        None
-    })
+    import Path._
+
+    val prefixPath: Path = {
+      val unslashed: Path = Path(prefix)
+      // Add slash automatically in the same way as Akka HTTP's `path` routing directive does.
+      unslashed match {
+        case p: Slash => p
+        case p => Slash(p)
+      }
+    }
+
+    def handlePath(req: HttpRequest): Option[Future[HttpResponse]] = {
+      @@tailrec
+      def handleRec(prefix0: Path, path0: Path): Option[Future[HttpResponse]] = (prefix0, path0) match {
+        case (Slash(prefix1), Slash(path1)) =>
+          println(s"$prefix0 is a prefix of $path0 - scanning subpath")
+          handleRec(prefix1, path1)
+        case (Segment(prefixSegment, prefix1), Segment(pathSegment, path1)) if prefixSegment == pathSegment =>
+          println(s"$prefix0 is a prefix of $path0 - scanning subpath")
+          handleRec(prefix1, path1)
+        case (_: SlashOrEmpty, Slash(Segment(method, Empty))) =>
+          println(s"$prefix0 is a prefix of $path0 - returning handler for $method")
+          Some(handle(req, method).recoverWith(GrpcExceptionHandler.default))
+        case _ =>
+          println(s"$prefix0 is NOT a prefix of $path0")
+          None
+      }
+      handleRec(prefixPath, req.uri.path)
+    }
```

Comment:
If we agree this is the way to go then we can factor some of this into a reusable method.

Reply:
Also is there some clever pattern match I can do to match a path prefix this instead of writing a custom function?